gms/gossiper.cc

```diff
@@ -2074,22 +2074,30 @@ void gossiper::build_seeds_list() {
     }
 }
 
-future<> gossiper::add_saved_endpoint(inet_address ep) {
+future<> gossiper::add_saved_endpoint(inet_address ep, permit_id pid) {
     if (ep == get_broadcast_address()) {
         logger.debug("Attempt to add self as saved endpoint");
         co_return;
     }
 
-    auto permit = co_await lock_endpoint(ep, null_permit_id);
+    auto permit = co_await lock_endpoint(ep, pid);
 
     //preserve any previously known, in-memory data about the endpoint (such as DC, RACK, and so on)
     auto ep_state = endpoint_state();
     auto es = get_endpoint_state_ptr(ep);
     if (es) {
+        if (es->get_heart_beat_state().get_generation()) {
+            logger.warn("Attempted to add saved endpoint after endpoint_state was already established with gossip: {}, at {}", es->get_heart_beat_state(), current_backtrace());
```

Comment:
> reduced on_internal_error_noexcept to warning (including a backtrace fwiw)

Why? The problem with `add_saved_endpoint` call from `topology_state_load` was fixed mainly because it complained loudly. Nobody will notice this `warn`-s.

Reply:
With `co_return` after this warning the issue is benign.
I have second thoughts about the internal error.
I think that the right direction would be to automatically restore the loaded state in `gossiper::reset_endpoint_state_map` as I described above, and then `gossiper::add_saved_endpoint` would either go away or become private in class gossiper.

Reply:
> With `co_return` after this warning the issue is benign.

But it still breaks the assumptions of `add_saved_endpoint — it shouldn't be called for existing ep, at least this is what is maintained in the current PR. Broken invariants should cause `inernal_error`.

I added `add_saved_endpoint` into `topology_state_load` some time ago, but now I don't like it. It incurs an overhead of grabbing the gossiper lock for every ep in the token_metadata each time `topology_state_load` is called, but `add_saved_endpoint` is only needed on node startup IIUC. Maybe we can do `add_saved_endpoint` in `topology_state_load` only once, on the first call.

Reply:
Don't we also need add_saved_endpoint() later than boot, if the node missed some topology changes and then catches up from group0 leader?

Reply:
This is discussed below...

Reply:
@tgrabiec please also take a look at the proposal here https://github.com/scylladb/scylladb/issues/16644